runtime/caml/mlvalues.h

```diff
@@ -84,13 +84,68 @@ typedef opcode_t * code_t;
 #define Unsigned_long_val(x) ((uintnat)(x) >> 1)
 #define Unsigned_int_val(x)  ((int) Unsigned_long_val(x))
 
-/* Encoded exceptional return values, when functions are suffixed with
-   _exn. Encoded exceptions are invalid values and must not be seen
-   by the garbage collector. */
+/* A 'result' type/monad for OCaml computations. */
+
+/* The [caml_result] type represents the result of computing an OCaml
+   term -- either a value or an exception:
+   - if [is_exception] is false,
+     the computation resulted in a value stored in [data]
+   - if [is_exception] is true,
+     the computation resulted in an exception stored in [data]
+
+   This is the C counterpart of a [('a, exn) result] type in
+   OCaml. Returning this type, instead of raising exceptions directly,
+   lets the caller implement proper cleanup and propagate the
+   exception themselves. */
+typedef struct {
+  _Bool is_exception;
```

Comment:
Note that C99 `_Bool` cannot be used in public headers for C++ compatibility as C++ doesn't define it. We chose not to include `stdbool.h` in public headers out of fear of compatibility problems, so we don't have access to `bool`. The current consensus is to use `int` for public booleans. See #12787, #12702 for previous discussions.

Reply:
It would be tempting to be able to use "`_Bool` or `bool`", because this plays nicely (nicer than `int`) with FFI based on code generation from the C headers. @gadmm tried Rust's `bindgen`, it creates a nice-looking Rust structure with a `bool` field.

However, in https://github.com/ocaml/ocaml/issues/12772#issuecomment-1824379144 @MisterDA hinted at ABI issues coming from the fact that `_Bool` and `bool` are not guaranteed to have the same size. This could be a deal-breaker, right?

I am unfamiliar with these aspects of language interoperability -- it is interesting to have to deal with the absence of a type of booleans in 2024 -- so I welcome guidance. My intuition would be to give up on the nice type and use `int` (or maybe `intnat`?) to avoid any potential issues.

Reply:
Use bool/_Bool or int; it should not make any change for Rust (nor to the caml_result type due to padding). I don't think ABI issues should be feared, as OCaml is unlikely to be supported on exotic platforms/compilers.